File: python/0904 FruitIntoBaskets.py
class Solution(object):
    def totalFruit(self, tree):
        """
        :type tree: List[int]
        :rtype: int
        """
        basket = collections.Counter()
        left = 0
        result = 0
        
        
        for right in range(len(tree)):
            basket[tree[right]]+=1
            
            while len(basket) > 2:
                basket[tree[left]]-=1
                if basket[tree[left]]<=0:
                    basket.pop(tree[left])
                left += 1
            
            result = max(result, sum(basket.values()))
            
        return result

File: javascript/0904 Fruit Into Baskets.js
/**
 * @param {number[]} tree
 * @return {number}
 */
var totalFruit = function(tree) {
    let best = 0;
    let left = 0;
    while(left<tree.length){
        const fruits={}
        let right = left;

        while(right<tree.length){
            if(!fruits[tree[right]]){
                if(Object.keys(fruits).length===2) break;
                else fruits[tree[right]] = 0;
            }
            fruits[tree[right]]++;

            right++;
        }

        best = Math.max(best, Object.values(fruits).reduce((a,b)=>a+b))
        if(right===tree.length) break;
        left++;
    }

    return best
};
